Read symbols from contract

HeeJu-K opened the pull request:
Option and future can select symbols read from the contract

The final goal is to read all symbols, filter them and show correct symbols on correct page (futures / options / powers), and use the symbol name to open positions with `trade`, `addMargin` and `removeMargin`. (please refer to the smart contract)

JoylimJY committed:
feat(develop_derivatives): refactor pool component

JoylimJY committed:
feat(develop_derivatives): refactor option component, finish trade function

ykchong45 committed:
fix: remove obsolete constantLoader, update reader addr

ykchong45 committed:
wip: remove duplicated Router address

JoylimJY committed:
fix(develop_derivatives): approve token & add margin

JoylimJY committed:
fix(develop_derivatives): add/rmv liquidity

JoylimJY committed:
fix(develop_derivatives): option component

HeeJu-K committed:
merge(develop_perpetual_hj):merged from develop_derivatives, when empty in options-pool, app crushes

HeeJu-K committed:
fix(develop_perpetual_hj): options-pool does not crush when input is empty and deletes 0 in the first digit

HeeJu-K committed:
Merge branch 'develop_perpetual' of https://github.com/ACY-Labs/acy-dex-interface into develop_perpetual_hj

HeeJu-K committed:
fix(develop_perpetual_hj): fixed option-pool-portfolio overview and alp datas, table data pending

HeeJu-K committed:
fix(develop_perpetual_hj):fixed option-pool-portfolio table data

HeeJu-K committed:
fix(develop_perpetual_hj):removed redundant codes in option-pool part, needs to be merged with develop_perpetual and develop_derivates

HeeJu-K committed:
merge(develop_perpetual_hj):merged with develop_perpetual, option page font is small, pending to merge with develop_derivatives

JoylimJY committed:
merge(develop_perpetual_hj): merge changes from main branch

HeeJu-K committed:
wip(develop_perpetual_heeju): option refactor delete repeating codes, style not changed and cannot close dropdown

HeeJu-K committed:
wip(develop_perpetual_heeju): can read option symbols from reader

HeeJu-K committed:
feat(develop_perpetual_heeju): can set option symbols, need change format of other pages

HeeJu-K committed:
merge(develop_perpetual_heeju): merged with develop_perpetuak

HeeJu-K committed:
feat(develop_perpetual_heeju): every page can select and see correct symbol

ykchong45 commented:
for now, connect to wallet then go to Option page would crash the site. Most probably due to the undefined result from symbol filtering with `.filter()`

HeeJu-K committed:
fix(develop_perpetual_heeju): option symbol fixed, commit to pull from develop_peretuak

HeeJu-K committed:
merge(develop_perpetual_heeju): merged with develop_pereptual

ykchong45 commented on src/pages/Future/Index.js:
```diff
@@ -9,34 +11,70 @@ import { getTokens, getContract } from '@/constants/powers.js';
+    fetcher: fetcher(library, Reader)
+  });
+  //future_tokens store every symbols in future and its data 
+  const future_tokens = symbolsInfo?.filter(ele=>ele[0] == "futures")
```
This chunk can be moved into a `useMemo()` to reduce the overhead of recalculation

ykchong45 committed:
fix: uncomment a line to fix undefined var

HeeJu-K committed:
fix(develop_perpetual): option future can buy and sell, deposit pending

HeeJu-K committed:
fix(develop_perpetual_heeju): deleted timeout in acycaurrencycard to resolve bug in deposit-token selector

HeeJu-K committed:
wip(develop_perpetual_heeju): derivative component now trade symbol with usd value; need more decimal displayed

HeeJu-K committed:
merge&fix(develop_perpetual_heeju):merged from develop_perpetual and can submit trade

HeeJu-K committed:
Merge branch 'develop_perpetual_heeju' of https://github.com/ACY-Labs/acy-dex-interface into develop_perpetual_heeju

ykchong45 commented on src/components/DerivativeComponent/index.js:
```diff
@@ -79,11 +83,12 @@ const DerivativeComponent = props => {
   const [marginToken, setMarginToken] = useState(tokens[1])
 
   const selectedTokenAmount = parseValue(selectedTokenValue, selectedToken && selectedToken.decimals)
```
The code doesn't compile on my end due to this duplicated line, I'll fix this.

ykchong45 reviewed (changes requested):
Please add new actions after switching symbols: 
1. close the side bar
2. update the chart

ykchong45 committed:
fix: remove duplicated variable definition

HeeJu-K committed:
fix(develop_perpetual_heeju): sidebar closed after token selection

HeeJu-K committed:
feat(develop_perpetual_heeju): fix to useMemo to store data read from symbolsInfo

ykchong45 committed:
fix: correct useMemo usage in fetching onchain symbols

ykchong45 committed:
fix: inconsistent hook invocations
in the old case, the 13th hook calls could be either undefined, or useEffect in line 162

ykchong45 reviewed (commented):
Recommend to compare the old code with new code for commit [803b7d3](https://github.com/ACY-Labs/acy-dex-interface/pull/98/commits/803b7d354a6df41617e196ac229eb7155d78c98b). A few critical points to take note:
1. `useMemo`'s syntax is similar to `useCallback`, not `useEffect`. 
2. `map` returns an array without explicitly pushing things to an array. In many cases `map` is more concise in syntax.
